[metricbeat] Add windows memory data to docker/memory

fearful-symmetry opened the pull request:
Leaving this as a draft since I'm not 100% sure how to handle vastly different stats across various operating systems. Will probably add more commits later as I tinker around. 

This addresses elastic/beats#11971 by adding the memory stats that are only available on windows. The change itself is fairly simple, but I have no practical experience with docker/go on windows, so I would appreciate help from any other folks. I added @narph to the reviewers list as well.

kaiyan-sheng commented on metricbeat/module/docker/memory/_meta/fields.yml:
```diff
@@ -4,6 +4,19 @@
   fields:
+    
+    - name: commit
+      type: long
```
`format: bytes` might be nice to add here.

jsoriano commented on metricbeat/module/docker/memory/data.go:
```diff
@@ -29,20 +29,31 @@ func eventsMapping(r mb.ReporterV2, memoryDataList []MemoryData) {
+				"pct":   memoryData.UsageP,
+				"max":   memoryData.MaxUsage,
+			},
+		}
```
I am fine with this approach if we cannot consider these metrics equivalent to others on Linux. The private working set could be similar to the RSS, and commit and commit peak could be similar to usage and max usage, but not sure.
Probably @narph or @andrewkroh can shed more light about this.
If they are equivalent or similar enough we should send them in the same fields.

narph replied:
Commit - memory space the application has reserved for use, so this will be allocated but not necessarily used. (This is composed of main memory (RAM) and disk (pagefiles).)
CommitPeak - the highest amount that the commit has reached since the operating system was last started or session.
PrivateWorkingSet - the amount of memory used by a process.
Not sure the memory stats are matching here so I would go for separate naming, we might want to look into that as well for the other memory related metricsets

jsoriano commented on metricbeat/module/docker/memory/data.go:
```diff
@@ -29,20 +29,31 @@ func eventsMapping(r mb.ReporterV2, memoryDataList []MemoryData) {
+		fields = common.MapStr{
+			"commit":              memoryData.Commit,
+			"commit_peak":         memoryData.CommitPeak,
+			"private_working_set": memoryData.PrivateWorkingSet,
```
Follow the structure of Linux metrics, in case some day docker supports limits in Windows and we can calculate percentages.

narph replied:
++

fearful-symmetry commented:
Adjusted things a tad so now we have commited bytes in `usage` at least.

Regarding @jsoriano 's comments: I agree that we should make these more consistent, and that things like `rss` on linux and Private Working Set on Windows are probably more or less equivalent. My concern is finding some sort of naming scheme that doesn't conflate the two, as _technically_ RSS is a linux concept, not a windows one. I worry that if we're not careful and we just cram metrics together, we might confuse people with linux-sounding Windows metrics and so on. I adjusted the format slightly so windows has `usage` now, but I would like input from some windows experts to see what else we can do to smooth out the naming.

jsoriano commented:
> I worry that if we're not careful and we just cram metrics together, we might confuse people with linux-sounding Windows metrics and so on.

Yeah, that's actually a point to discuss. The good thing of having all the equivalent metrics in the same fields is that you can build dashboards or queries showing this information together more easily. One option can be to keep OS-specific fields, but also duplicate these values into common fields. This could be added afterwars in any case.

fearful-symmetry commented:
> One option can be to keep OS-specific fields, but also duplicate these values into common fields. This could be added afterwars in any case.

That's one option, yah! We did something similar with the RAID metricset, where we had a detail section, and a normalized section for RAID states.

narph commented:
@fearful-symmetry , I have not commented on the initial naming as I found it to be clear enough.
```
"commit":              memoryData.Commit,
"commit_peak":         memoryData.CommitPeak,
"private_working_set": memoryData.PrivateWorkingSet,
```
You could go for `Commit size` instead of `Commit` here.

I would stir away from `usage` or `memory usage` when talking about private working set as there are more types here (of memory usage): working set, private working set, working set will also include the memory shared by other processes.
https://answers.microsoft.com/en-us/windows/forum/windows_10-performance/physical-and-virtual-memory-in-windows-10/e36fb5bc-9ac8-49af-951c-e7d39b979938
http://blogs.microsoft.co.il/sasha/2016/01/05/windows-process-memory-usage-demystified/

@jsoriano, Is there a way we can store only the os specific information and utilize the field alias to create a common visualization/dashboard for the user instead of going for duplication? (I have not tried this)

fearful-symmetry commented:
Thanks for your added input @narph. 
I definitely like the alias idea, but I have absolutely no idea how to do that.

jsoriano commented:
> @jsoriano, Is there a way we can store only the os specific information and utilize the field alias to create a common visualization/dashboard for the user instead of going for duplication? (I have not tried this)

I don't think we can use the alias field for this. Aliases are defined in the index mapping (not per document), and can only be defined in a 1 to 1 way, so we couldn't create an alias pointing to several OS-specific metrics.

I'd leave both OS-specific metrics by now as they are different things (thanks @narph for the clarifications!), and we can evaluate later what to do if we create visualizations for this. Maybe this can be solved in the visualization or the dashboard itself.

fearful-symmetry commented:
> I'd keep both OS-specific metrics in the events by now 
@jsoriano So we're good as-is? Should I take this out of draft?

jsoriano commented:
> > I'd keep both OS-specific metrics in the events by now
> 
> @jsoriano So we're good as-is? Should I take this out of draft?

Yeah, I think this can go for review. Only let me know what you think about this comment https://github.com/elastic/beats/pull/12172/files#r283109849

It would be to change the metrics to something like this:
```
			"usage": common.MapStr{
				"commit": common.MapStr{
                                    "total": memoryData.Commit,
				    "peak":  memoryData.CommitPeak,
                                },
			},
			"private_working_set": common.MapStr{
                            "total": memoryData.PrivateWorkingSet,
                        },
```

In case some day we add also percentages or other aggregations.

fearful-symmetry commented:
Ahh, alright, I wasn't 100% sure what you meant by that, so I figured the `usage` dict was what you meant. Yah, I like that idea.

fearful-symmetry commented:
Okay, I think I'm confusing myself with the fields.yml format, but it should be good now.

jsoriano commented:
@fearful-symmetry are you going to do the change for commit and commit peak too?

> It would be to change the metrics to something like this:
> 
> ```
> 			"usage": common.MapStr{
> 				"commit": common.MapStr{
>                                     "total": memoryData.Commit,
> 				    "peak":  memoryData.CommitPeak,
>                                 },
> 			},
> ```

fearful-symmetry commented:
Oh, is that how we want to do it? I figured we could just use the existing `pct` type in the `commit` object.

fearful-symmetry commented:
Okay. Hopefully this makes sense now.

jsoriano commented on metricbeat/module/docker/memory/_meta/fields.yml:
```diff
@@ -34,6 +53,11 @@
+          type: long
+          format: bytes
+          description: >
+            Peak committed bytes on Windows
```
This is not needed anymore.

jsoriano commented on metricbeat/module/docker/memory/_meta/fields.yml:
```diff
@@ -4,7 +4,26 @@
+          type: long
+          format: bytes
+          description: >
+            peak bytes
```
```suggestion
            Peak bytes
```
